app/server/MergedServer.ts

```diff
@@ -0,0 +1,234 @@
+    ms.flexServer.addHosts();
+
+    ms.flexServer.addDocWorkerMap();
+
+    if (ms.hasComponent("home") || ms.hasComponent("static")) {
+      await ms.flexServer.addAssetsForPlugins();
+    }
+
+    if (ms.hasComponent("home")) {
+      ms.flexServer.addEarlyWebhooks();
+    }
+
+    if (ms.hasComponent("home") || ms.hasComponent("docs") || ms.hasComponent("app")) {
+      ms.flexServer.addSessions();
+    }
+
+    ms.flexServer.addAccessMiddleware();
+    ms.flexServer.addApiMiddleware();
+    await ms.flexServer.addBillingMiddleware();
+
+    return ms;
+  }
+
+  public readonly flexServer: FlexServer;
+  private readonly _serverTypes: ServerType[];
+  private readonly _options: ServerOptions;
+
+  private constructor(port: number, serverTypes: ServerType[], options: ServerOptions = {}) {
+    this._serverTypes = serverTypes;
+    this._options = options;
+    this.flexServer = new FlexServer(port, `server(${serverTypes.join(",")})`, options);
+  }
+
+  public hasComponent(serverType: ServerType) {
+    return this._serverTypes.includes(serverType);
+  }
+
+
+  public async run() {
+
+    try {
+      await this.flexServer.start();
+
+      if (this.hasComponent("home")) {
+        this.flexServer.addUsage();
+        if (!this.hasComponent("docs")) {
+          this.flexServer.addDocApiForwarder();
+        }
+        this.flexServer.addJsonSupport();
+        this.flexServer.addUpdatesCheck();
+        await this.flexServer.addLandingPages();
+        // todo: add support for home api to standalone app
+        this.flexServer.addHomeApi();
+        this.flexServer.addBillingApi();
+        this.flexServer.addNotifier();
+        await this.flexServer.addTelemetry();
+        await this.flexServer.addHousekeeper();
+        await this.flexServer.addLoginRoutes();
+        this.flexServer.addAccountPage();
+        this.flexServer.addBillingPages();
+        this.flexServer.addWelcomePaths();
+        this.flexServer.addLogEndpoint();
+        this.flexServer.addGoogleAuthEndpoint();
+        this.flexServer.addInstallEndpoints();
+      }
+
+      if (this.hasComponent("docs")) {
+        this.flexServer.addJsonSupport();
+        await this.flexServer.addTelemetry();
+        await this.flexServer.addDoc();
+      }
+
+      if (this.hasComponent("home")) {
+        this.flexServer.addClientSecrets();
+      }
+
+      this.flexServer.finalizeEndpoints();
+      await this.flexServer.finalizePlugins(this.hasComponent("home") ? checkUserContentPort() : null);
+      this.flexServer.checkOptionCombinations();
+      this.flexServer.summary();
+      this.flexServer.ready();
+
+      // Some tests have their timing perturbed by having this earlier
+      // TODO: update those tests.
+      if (this.hasComponent("docs")) {
+        await this.flexServer.checkSandbox();
+      }
+    } catch(e) {
+      await this.flexServer.close();
+      throw e;
+    }
+  }
+}
+
+export async function startMain() {
```

Comment:
Can you remind me of the motivation for this change was again? It is hard to pick out the changes.

Reply:
There was a need to modify certain aspects of the "merged server" before starting it, so I broke the `mergedServerMain` function down into two parts - create then run. Technically this is no longer needed since I started using ICreate for the DocStorageManager decorator, but I think it's good to keep. The original function does too much.

Reply:
I'm sorry to ask this, but if you don't need it, can you revert it? The original doesn't look bad to me, and reviewing this change is taking me some effort to line things up and understand them again.

Reply:
For example, figuring out the deal with `setLoginSystem`

Reply:
`setLoginSystem` is about to be deprecated in favor of `ICreate.getLoginSystem`, so removed here.

Reply:
Reverting isn't hard, but I'd like to defend this change - not a strong preference, more of a suggestion.

I'm currently using this workflow:
- Create MergedServer and run it in the background
- Extract FlexServer from MergedServer and initialize Desktop components in the meantime
- Wait for both electron and the MergedServer to finish initialization

This maximizes asynchrony to make startup fast, but relies on 2-step MergedServer creation. (Just double checked and I'm doing it wrong - sorry! I should have obtained the MergedServer Promise before proceeding with Desktop initialization. So the code as of now doesn't actually match what I said, but it's an easy fix.)

I think this change will likely be desirable someday, and it's not a huge change. The only consequence is that all
```typescript
const flexServer = await mergedServerMain(port, serverTypes);
```
will now need to be replaced with
```typescript
const mergedServer = await MergedServer.create(port, serverTypes);
const mergedServerInitialized = mergedServer.run();
const flexServer = mergedServer.flexServer;
await mergedServerInitialized; // feel free to `await` later, or use it in Promise.all()
```
and now there is freedom to do something else while the FlexServer starts.

Do we really want to revert?

Reply:
Update: As discussed offline, this change cannot be reverted.

The FlexServer's DocStorageManager must be decorated after the DocRegistry is initialized. The DocStorageManager is created in the `addDoc()` method, invoked by `mergedServerMain.ts` after `flexServer.start()`. This means DocRegistry initialization must happen after the MergedServer is created but before it calls `flexServer.addDoc()`. To avoid having electron-specific code in grist-core, this is not feasible without breaking MergedServer creation into two stages.

Reply:
>So you need the separation of create() and run() in desktop, is that correct?
Yes.

I left some comments on changes I made as per your request. Nothing major except that the login system is now handled via `ICreate` so removed from here. Marking this as resolved.